На проверку "Хеллоу Ворлд"

gderpy opened the pull request:
Отработаны основные функции админ-панели/Авторизация пользователя

1. Добавлены модели: Article (Статьи), Category (Категории)
2. Пользовательское поле
3. Панель поиска, панель фильтрации
4. Добавлена авторизация пользователя

UPD 02.09.2024:

5. Исправлены миграции
6. Авторизация
7. Регистрация
8. Страница пользователя
9. Функция редактирования данных профиля

UPD 03.09.2024:

10. Пришлось снести все миграции, т.к. возникал конфликт из-за индекса. Но в целом они и не были особо нужны. 
11. Реализовал лайки
12. Приделал HTMX

gderpy committed:
Добавлена админ-панель

gderpy committed:
Добавлена админ-панель

gderpy committed:
Merge branch 'feature/admin-panel' into develop

gderpy committed:
Первоначальные настройки авторизации пользователей

gderpy committed:
При вводе данных в процессе авторизации происходит валидация, пользователя перекидывает на главную страницу сайта

gderpy committed:
Прописан выход из профиля с переходом на ссылку авторизации

gderpy committed:
После авторизации будет отображено имя пользователя и кнопка 'выйти'

gderpy committed:
Реализована авторизация и возможность выхода из аккаунта

gderpy committed:
При попытке добавить статью неавторизованному пользователю - будет перекидывать на страницу авторизации

gderpy committed:
Добавлено отображение автора статьи

gderpy committed:
Добавлена первоначальная регистрация пользователя
[Реализовано через функция представления]

gderpy committed:
Переделал регистрацию через класс представления - RegisterUser(CreateView)

gderpy committed:
Доработана регистрация пользователей
1) Добавлена дополнительная модель UserProfile, обеспечивающая дополнительные поля к основной модели User
2) Задана логика для админ-панели для правильного отображения двух связанных моделей и их данных
3) Прописаны сигналы для передачи в модель UserProfile модели User для связки 1:1
4) Добавлена логика в класс представления RegisterUser, связанная с доп моделью UserProfile

gderpy committed:
Решаем проблему с редактированием данных пользователя

gderpy committed:
Добавлена возможность редактирования данных пользователя

gderpy committed:
Добавлены стили Bootstrap в редактирование профиля

fallenflint commented on main/admin.py:
```diff
@@ -1,3 +1,38 @@
+        self.message_user(request, f"Снято с публикации {count} записей.", messages.WARNING)
+
+
+# admin.site.register(Article, ArticleAdmin)
```
Лучше сразу убирай комменты - это только кажется, что оно надо, на самом деле не надо.

fallenflint commented on main/migrations/0011_merge_20240811_2030.py:
```diff
@@ -0,0 +1,14 @@
+        ('main', '0010_remove_category_parent'),
+    ]
+
+    operations = [
```
Эта миграция выглядит странно. Она ничего не делает же?
Я бы предложил из 0012 сюда всё перетащить, а саму 12ю удалить.
(главное перетащить и dependencies и operations)

fallenflint commented on main/migrations/0012_alter_article_options_alter_category_options_and_more.py:
```diff
@@ -0,0 +1,69 @@
+
+    dependencies = [
+        ('main', '0011_merge_20240811_2030'),
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
```
Вот эту строчку (12ю) можно перенести в миграцию 0011

fallenflint commented on main/migrations/0012_alter_article_options_alter_category_options_and_more.py:
```diff
@@ -0,0 +1,69 @@
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+    ]
+
+    operations = [
```
И вот это вот всё (внутри operations)

fallenflint commented on main/templates/main/index.html:
```diff
@@ -7,7 +7,7 @@
             <div class="col-12 d-flex gap-2">
                 <div class="avatar"><i class="fa-solid fa-user-ninja"></i></div>
-                <div class="article-author">Darry Poster</div>
+                <div class="article-author">{{ article.author.username|default:"Аноним" }}</div>
```
Как заглушка на случай отсутствия - хорошо, молодец, но вообще с точки зрения бизнес-логики - ты вряд ли хочешь держать анонимные статьи на своём блоге :) А если захочешь - то всегда успеешь добавить эту логику.
Я бы предложил убрать null=True на авторе.
Такие миграции делаются в 2 шага:
1. заполняешь все Article.author (в миграции прямо можно сделать RunPython) https://realpython.com/data-migrations/
2. Делаешь null=False (просто убираешь null=True).

fallenflint commented on templates/base.html:
```diff
@@ -47,8 +47,8 @@
-                            <a href="{% url 'login' %}" class="text-white text-decoration-none px-3 py-1 rounded-2">Войти</a>
-                            <a href="{% url 'registration' %}" class="text-black text-decoration-none px-3 py-1 rounded-2" 
+                            <a href="{% url 'users:login' %}" class="text-white text-decoration-none px-3 py-1 rounded-2">Войти</a>
+                            <a href="{% url 'users:logout' %}" class="text-black text-decoration-none px-3 py-1 rounded-2" 
```
Кажись, здесь должен быть не логаут :) (а registration)

fallenflint commented on templates/base.html:
```diff
@@ -65,8 +65,16 @@
+                            </form>
+                        {% else %}
+                            <a href="{% url 'users:login' %}" class="btn btn-outline-dark">Войти</a>
+                            <a href="{% url 'users:logout' %}" class="btn btn-dark">Регистрация</a>
```
И здесь тоже

fallenflint commented on templates/base_2.html:
```diff
@@ -27,8 +27,13 @@
+                            <a href="{% url 'users:logout' %}" class="btn btn-outline-dark ms-3">Выйти</a>
+                            {% else %}
+                            <a href="{% url 'users:login' %}" class="btn btn-outline-light">Войти</a>
+                            <a href="{% url 'users:logout' %}" class="btn btn-secondary">Регистрация</a>
```
И здесь

fallenflint commented on templates/base_2.html:
А в целом, в чем назначение base_2?

gderpy replied:
base.html - содержит в себе именно основное навигационное меню: с категориями, строкой поиска.
![Снимок экрана от 2024-08-24 22-23-07](https://github.com/user-attachments/assets/903e36c6-a694-4b10-a4fe-fc61d100dc20)

base_2.html - оно больше второстепенное, например, когда пользователь заходит почитать статью, то меню уже будет в упрощенном виде.
![Снимок экрана от 2024-08-24 22-23-39](https://github.com/user-attachments/assets/7d699f5e-a1f9-44ab-91db-a4303eee6b47)

fallenflint replied:
тут обычно делают 1 base, а дальше расщепляют его на layout'ы например.
Просто принято считать что base - это прям самая главная основа, корень.
В принципе, правило не железное, но тогда вопрос к неймингу - из имен новый разработчик твоего проекта вряд ли поймёт, в чем различие. Отсюда общее "правило" хорошего нейминга - это назвать так, чтобы было понятно, что это, зачем это.

fallenflint replied:
И да, хороший нейминг можно годами оттачивать :)

fallenflint commented on users/tests.py:
фалы tests.py и вобоще все, что пустые или содержат заглушки (ничего не делающие) - смело удаляй.
Если тесты начнёшь писать - то там лучше немного иначе всё организовать, не стандартным вариантом джанги.

fallenflint commented on users/views.py:
```diff
@@ -0,0 +1,42 @@
+from django.contrib.auth.forms import AuthenticationForm
+from .forms import LoginUserForm
+
+# Create your views here.
```
В этом фале тоже можно всё закомменченное удалять.

fallenflint commented on users/views.py:
```diff
@@ -0,0 +1,42 @@
+#     }
+#     return render(request, "users/login.html", context=data)
+
+class LoginUser(LoginView):
```
Обычно в джанго-мире принято вьюшки обзывать с View на конце, так тебя быстрее/проще поймут коллеги.

fallenflint commented on users/views.py:
```diff
@@ -0,0 +1,42 @@
+    form_class = LoginUserForm
+    template_name = "users/login.html"
+    extra_context = {
+        "title": "Авторизация",
```
Вот эта штука в данном случае похожа на нарушение паттерна MVC, когда у нас логика от представления отделена. Лучше в самом шаблоне base.html вместо
```
<title>{{ title }}</title>
# сделать
<title>{% block title %}{{ title }}{% endblock %}</title>
```
И в таком случае ты в тех местах, где у тебя title берется из модельки - можешь подставлять его как раньше, а в любом другом месте, например в login.html - просто перебить блок title.

fallenflint commented on users/views.py:
```diff
@@ -0,0 +1,42 @@
+        "title": "Авторизация",
+    }
+
+    def get_success_url(self):
```
Эту штуку ещё можно базовыми настройками реализовать. https://docs.djangoproject.com/en/5.1/ref/settings/#login-redirect-url

LOGOUT_REDIRECT_URL = reverse_lazy("home")

fallenflint commented on main/views.py:
```diff
@@ -36,22 +40,11 @@ def write_article(request):
-
 def show_category(request, slug_name):
     category = get_object_or_404(Category, slug=slug_name)
     articles = Article.published.filter(category_id=category.pk)
```
Кстати, здесь можно упростить запрос:
```
articles = get_list_or_404(Article.published, category__slug=slug_name)
# и после этого предыдущая строчка вообще не нужна
```

Дока: https://docs.djangoproject.com/en/5.1/topics/http/shortcuts/#get-list-or-404

gderpy committed:
Исправлены некоторые ошибки
- Подправлены миграции

gderpy committed:
Добавлен categories.json, в котором хранятся первоначальные данные по категориям

gderpy committed:
Доработаны некоторые мелкие моменты на сайте
1) Немного подправлена форма регистрации
2) После отправки статьи пользователем, его переносит в шаблон, уведомляющем, что статьи на рассмотрении модерации

gderpy committed:
Добавлены лайки
1) Добавлена фукнция лайков
2) Прикрепил HTMX
3) Пришлось удалить все миграции из-за конфликтного индекса. Ну там и не было ничего нужного

fallenflint commented on .gitignore:
```diff
@@ -4,4 +4,5 @@ db.sqlite3
 *.drawio
-ubcategory.name }}
\ No newline at end of file
+ubcategory.name }}
```
Кажется, сюда какая-то бяка попала)

fallenflint commented on main/migrations/0012_merge_20240902_1909.py:
Вот эта миграция, само её существование - маленький косячок. Значит, предыдущие миграции разделились в 2 ветки

fallenflint commented on users/admin.py:
```diff
@@ -1,3 +1,41 @@
+class CustomUserAdmin(UserAdmin):
+    inlines = (UserProfileInline, )
+    list_display = (
+        "username", "email", "first_name", "last_name", "get_sex", "get_birthdate"
```
Смотри в чём фишка, сразу лучше на такое обращу внимание:
Ты тут в list_display указываешь get_sex и get_birthdate. Запомним (остальное - в следующем комменте)

fallenflint commented on users/admin.py:
```diff
@@ -1,3 +1,41 @@
+    )
+
+    def get_sex(self, obj):
+        return obj.userprofile.sex
```
А вот здесь ты идёшь в профиль.
Что будет делать джанга, когда ты запросишь страницу юзеров? Она выберет юзеров, примерно так:
queryset = CustomUser.objects.all()

Дальше условно будет:
for user in queryset: ... и в какой-то момент пойдёт обращение к user.get_sex-> user.userprofile.sex
Что сделает джанга? Она на каждого юзера начнёт делать по отдельному запросу к таблице с профилями.
И у тебя будет O(N) запросов в БД. А должно быть O(1), т.е. не важно скольколь у тебя пользователй, запросы не должны расти линейно. 1 юзер в бд - не более 10 примерно запросов из админки. 10000 юзеров в бд - всё равно не более 10 запросов нужно держать.

Что делать? надо научить админку в запросе к CustomUser выгребать ещё и 2 поля из профиля. Для этого в джанге есть select_related: https://docs.djangoproject.com/en/5.1/ref/models/querysets/#select-related
CustomUser.objects.all().select_related('userprofile')

Ну и следующий вопрос - а куда это писать?
Есть у админ-класса такой метод https://docs.djangoproject.com/en/5.1/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_queryset
Ты его переопределяешь и добавляешь seleсt_related. примерно так:
def get_queryset(self, request):
    return super().get_queryset(request).select_related('userprofile')

fallenflint commented on users/forms.py:
```diff
@@ -12,4 +15,118 @@ class LoginUserForm(AuthenticationForm):
+        max_length=50,
+        required=True, 
+        label="Имя",
+        widget=forms.TextInput(attrs={"class": "form-control", "id": "inputFirstName"}))
```
по поводу вот этого подхода - всё норм, так делают, но я добавлю комментарий.
Имеет смысл так делать, когда у тебя мало шаблонов, а форм дохуя и потому ты делаешь 1 условно form.html, и в него пихаешь разные формы.
В вебе, т.е. в сайтах, как правило ситуация обратная - 1 форма - 1 странчка где она используется, и формы/странички рисует фронтендер/дизайнер. То есть это он определяет, как будет всё выглядеть. А ты уже правишь формы под это дело. Потому я топлю за подход в простом совецком вебе - делать все css-айдишники, классы - прям в html, т.е. у нас за визуал отвечает html, живёт отдельно. А форма - она за поведение - тип данных, ограничения/валидации и всё такое.
